### PR TITLE
[BIOMAGE-1049]Implement functionality to upload files for incomplete samples

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Background
#### Link to issue 
https://biomage.atlassian.net/secure/RapidBoard.jspa?rapidView=1&modal=detail&selectedIssue=BIOMAGE-1049
#### Link to staging deployment URL 
https://ui-stefan-ui325.scp-staging.biomage.net/
#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/components/data-management/UploadDetailsModal.jsx [hunk 1]
```diff
@@ -99,7 +99,7 @@ const UploadDetailsModal = (props) => {
         }}
         style={{ width: '140px', marginBottom: '10px' }}
       >
-        Replace file
+        {bundleName ? 'Replace file' : 'Upload'}
```
Comment (robioson) [inline src/components/data-management/UploadDetailsModal.jsx]: You could add a comment explaining that `bundleName` will be undefined if the file hasn't been uploaded

File: src/utils/checkIfFileValid.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    techOptions[selectedTech].validMimeTypes
+      .includes(
+        mime.lookup(fileName),
+      )
+    || techOptions[selectedTech].validExtensionTypes
+      .includes(
+        path.extname(fileName),
+      )
+  );
+  const acceptedFilesRegexp = `(${techOptions[selectedTech].acceptedFiles.join('|')})$`;
```
Comment (robioson) [inline src/utils/checkIfFileValid.js]: I know that naively writing unions using `join('|')` can cause issues with escaping. Might not be an issue here but Ruby has `Regexp.union` for this, I wonder if we could do something similar.

File: src/components/data-management/ProjectDetails.jsx [hunk 3]
```diff
@@ -551,15 +534,14 @@ const ProjectDetails = ({ width, height }) => {
     if (!uploadDetailsModalDataRef.current) {
       return;
     }
-
     const { sampleUuid, file } = uploadDetailsModalDataRef.current;
-
-    const bucketKey = `${activeProjectUuid}/${sampleUuid}/${file.name}`;
+    const name = file.name || bundleToUpload.name;
```
Comment (robioson) [inline src/components/data-management/ProjectDetails.jsx:537]: Similar to below review comment, a comment here would be good to explain when each path is taken.

**Review by robioson — COMMENTED** [review]
Looks good, few minor comments

File: src/components/data-management/UploadDetailsModal.jsx [hunk 4]
```diff
@@ -139,7 +140,7 @@ const UploadDetailsModal = (props) => {
         {!isSuccessModal
           && (
             <Row style={{ marginTop: '5px', marginBottom: '5px' }}>
-              The following file has failed to upload
+              The following file has failed to upload or was not uploaded
```
Comment (cosa65) [inline src/components/data-management/UploadDetailsModal.jsx]: I think we should change this message to depend on the status that the modal is filled with currently.
"The following file has failed to upload" for UPLOAD_ERROR
"The following file was not uploaded" for FILE_NOT_FOUND

File: src/components/data-management/UploadDetailsModal.jsx [hunk 5]
```diff
@@ -127,7 +127,8 @@ const UploadDetailsModal = (props) => {
       footer={(
         <Row style={{ width: '100%', justifyContent: 'center' }}>
           <Col>
-            {isSuccessModal ? downloadButton() : retryButton()}
+            {/* render retry button only if file was tried to be uploaded */}
+            {bundleName && (isSuccessModal ? downloadButton() : retryButton())}
```
Comment (cosa65) [inline src/components/data-management/UploadDetailsModal.jsx]: A more strict way to check for this (as in it fails if we reach an invalid state) could be checking that the status is not FILE_NOT_FOUND instead of checking for bundleName to exist

File: src/components/data-management/UploadDetailsModal.jsx [hunk 6]
```diff
@@ -150,10 +151,12 @@ const UploadDetailsModal = (props) => {
           <Col span={5}>Category</Col>
           <Col span={10}>{fileCategory}</Col>
         </Row>
-        <Row style={{ marginTop: '5px', marginBottom: '5px' }}>
-          <Col span={5}>Filename</Col>
-          <Col span={10}>{bundleName}</Col>
-        </Row>
+        {bundleName && (
```
Comment (cosa65) [inline src/components/data-management/UploadDetailsModal.jsx]: perhaps use FILE_NOT_FOUND instead as in previous comment?

File: src/redux/actions/samples/updateSampleFile.js [hunk 7]
```diff
@@ -16,18 +17,22 @@ const updateSampleFile = (
 ) => async (dispatch, getState) => {
   const updatedAt = moment().toISOString();
   const sample = getState().samples[sampleUuid];
+  // we'll need to remove the hard-coded 10x tech type once we start
+  // supporting other types and save the chosen tech type in redux
+  const valid = checkIfFileValid(fileName, '10X Chromium');
```
Comment (cosa65) [inline src/redux/actions/samples/updateSampleFile.js]: we are using `updateSampleFile` only in `processUpload` and usually not to create a new sample file but to update an existing one so it is usually not necessary to do this check, I assume this validity check was added here so that it is run when we are using `UploadDetailsModal` to update a file, could it be moved into that modal instead then?

File: src/utils/checkIfFileValid.js (hunk 2 above)
Comment (StefanBabukov) [reply src/utils/checkIfFileValid.js]: what do you mean i didnt get it - .join just adds the character between every element in the array. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join Or you mean something else?

File: src/redux/actions/samples/updateSampleFile.js [hunk 8]
```diff
@@ -16,18 +17,22 @@ const updateSampleFile = (
...
 
   try {
     // Save sample only if upload is successful or error
     if (fileDiff.upload.status === UploadStatus.UPLOADED
```
Comment (cosa65) [inline src/redux/actions/samples/updateSampleFile.js]: this is not part fo your PR, but could you perhaps change this if for `[UPLOADED, UPLOAD_ERROR].includes(fileDiff.upload.status)`

File: src/redux/actions/samples/updateSampleFile.js [hunk 9]
```diff
@@ -16,18 +17,22 @@ const updateSampleFile = (
...
       || fileDiff.upload.status === UploadStatus.UPLOAD_ERROR) {
       const diffObject = {
-        fileNames: sample.fileNames.add(fileName),
+        fileNames: sample.fileNames,
```
Comment (cosa65) [inline src/redux/actions/samples/updateSampleFile.js]: why was this line changed? If filenames is still a set then it should not be a problem, right?

File: src/redux/reducers/samples/samplesFileUpdate.js [hunk 10]
```diff
@@ -20,8 +20,7 @@ const samplesFileUpdate = (state, action) => {
 
   newFile.lastModified = lastModified;
 
-  const newFileNames = new Set([...state[sampleUuid].fileNames, fileName]);
-
+  const newFileNames = Array.from(new Set([...state[sampleUuid].fileNames, fileName]));
```
Comment (cosa65) [inline src/redux/reducers/samples/samplesFileUpdate.js:23]: In the initial state fileNames is a set, why are we doing this change?

File: src/utils/checkIfFileValid.js (hunk 2 above)
Comment (robioson) [reply src/utils/checkIfFileValid.js]: What I mean is that the things you are joining are strings being turned into a regex. However the strings (file names) contain dots, which have a special meaning in regex, so technically need to be escaped. Sorry, I know it's weird to use another language as an example but you will see that `Regexp.union` safely escapes special characters:

```irb
irb(main):012:0> Regexp.union(*acceptedFiles)
=> /barcodes\.tsv|barcodes\.tsv\.gz|features\.tsv|features\.tsv\.gz|genes\.tsv|genes\.tsv\.gz|matrix\.mtx|matrix\.mtx\.gz/
irb(main):013:0> acceptedFiles.join('|')
=> "barcodes.tsv|barcodes.tsv.gz|features.tsv|features.tsv.gz|genes.tsv|genes.tsv.gz|matrix.mtx|matrix.mtx.gz"
```
Because dots mean "match any one character", my prediction is that the current logic will allow any characters in place of the dots. This is easy to test.

Also, I know you did not write this and just moved it, but I just noticed it here!

Comment (robioson) [reply src/utils/checkIfFileValid.js:19]: OK, I think the mime type check prevents this from ever actually being an issue but it is technically wrong.

Comment (robioson) [reply src/utils/checkIfFileValid.js]: We could use https://stackoverflow.com/a/29700268/7799676.

```js
const acceptedFileRegexp =  `(${techOptions[selectedTech].acceptedFiles.map(_.escapeRegExp).join('|')})$`
```

Seems to be correct

Comment (robioson) [reply src/utils/checkIfFileValid.js]: image
Yeah, it is actually possible to witness this on the platform (the upload fails but beforehand it is marked as valid). My fix should work I think.

File: src/redux/actions/samples/updateSampleFile.js (hunk 9 above)
Comment (StefanBabukov) [reply src/redux/actions/samples/updateSampleFile.js]: Relates to the below comment and also sample.fileNames was already updated to include the new file

File: src/redux/reducers/samples/samplesFileUpdate.js (hunk 10 above)
Comment (StefanBabukov) [reply src/redux/reducers/samples/samplesFileUpdate.js:23]: The API expects an array not a set and I was getting errors.

File: src/redux/actions/samples/updateSampleFile.js (hunk 7 above)
Comment (StefanBabukov) [reply src/redux/actions/samples/updateSampleFile.js]: But we still need to provide a `valid` field here so that the analysis can be launched. Do you think it should be just valid: true here, while the check is done in the `UploadDetailsModal` and if valid is true, then call `updateSampleFile`?

File: src/utils/checkIfFileValid.js (hunk 2 above)
Comment (StefanBabukov) [reply src/utils/checkIfFileValid.js]: Thanks for the explanation! I changed it.

File: src/redux/actions/samples/updateSampleFile.js (hunk 9 above)
Comment (cosa65) [reply src/redux/actions/samples/updateSampleFile.js]: Ohh right, so it is already added in the reducer which means this is unnecessary.

Looks good and I definitely think it is better to perform all the setup for updates in the reducer while the action just sends what it received (and/or performs api calls), however I think I would go further and remove the `fileNames` field from this diff.

Comment (cosa65) [reply src/redux/actions/samples/updateSampleFile.js]: Also something else that we will need to do eventually is separate sampleFile updates from sampleFile posts to the api a little bit more, but that's for a different task.

File: src/components/data-management/UploadDetailsModal.jsx [hunk 11]
```diff
@@ -37,6 +37,7 @@ const UploadDetailsModal = (props) => {
   }, [replacementFileBundle]);
 
   const isSuccessModal = status === UploadStatus.UPLOADED;
+  const isNotUploaded = status === UploadStatus.FILE_NOT_FOUND;
```
Comment (cosa65) [inline src/components/data-management/UploadDetailsModal.jsx]: could it be isNotUploadedModal instead? to be more coherent with the uploaded status check in line 39

File: src/redux/actions/samples/updateSampleFile.js (hunk 7 above)
Comment (cosa65) [reply src/redux/actions/samples/updateSampleFile.js]: Yes, I believe that's the pattern we are already following in FileUploadModal, and apart from this we should use the same method for both modals (like the utils function you are using, perhaps compare it with your implementation and check if they can be integrated).

**Review by cosa65 — APPROVED** [review]
